코드리뷰 여섯번째

Mangcho opened the pull request:
안녕하세요 선배님 ERD는 현재 서비스 브랜치에 존재하는 ERD가 아니라 마스터에 존재하는 아래 이 사진을 참조해주시면 감사하겠습니다.
![ERD](https://user-images.githubusercontent.com/46319374/197136999-bff55ff6-5208-4f9b-a78a-68d7e3643c46.jpg)

먼저 이전 코드리뷰에서 적용하지 않은 것은 다음 부분입니다.
```text 
유저를 찾는 서비스 로직은 다른 데에서도 많이 쓰일 겁니다. 
그러면 그냥 유저를 내보내는 게, 추후 재사용성이 높을 거에요. 
그리고 그렇게 수정할 거라면, 유저를 찾을 때는 항상, 
password와 같이 외부로 노출되면 안 되는 값에 대한 필터링을 해주세요.
```

아직 이해를 못했습니다.. 따로 이번 프로젝트에서는 그렇게 해야할 필요도 느끼지 못했구요

질문하고 싶은게 있는데
현재 session 사용 중에 서비스와 라우팅의 분리대로라면 logout을 어떻게 구현할지 의문이 있습니다.   **req.session.destory** 를 사용하지 못할것 같은데 그냥 제가 직접 DB에 해당 세션 기록을 삭제하라고 날려도 괜찮을까요?
애러처리는 지금 console.log로 대체하고 있는데 애러는 보통 어떻게 처리하는지 궁금합니다.

Mangcho committed:
추가 : 신규 모듈 설치, ESM 코드로 변경
1. express-validator 설치
2. type: module 추가로 ESM 코드로 인식하게 변경

Mangcho committed:
추가 : 값 validation 추가 및 기능 추가
1. 기존에 존재하던 모든 auth 로직 empty 및 타입 검증 추가
2. 회원가입시 사용하는 imei값 검증 기능 추가

Mangcho committed:
수정, 추가 : imei 검증 기능 추가 및 기존 기능 로직 일부 변경
1. CheckImei 기능 추가
2. #7 코드리뷰 지적 반영하여 DB create 문 사용시 생성불가는 error로 처리되므로 return 코드 변경

Mangcho committed:
보완 : config.js sequelize.js default export 변경
default export에서 named export로 변경

Mangcho committed:
보완 : ORM 및 DB 커넥터 설정파일 타임존 및 기본 설정 추가
#7 코드리뷰 지적 반영하여 기본 설정 추가
sequelize 타임존 관련 설정이 없길래 타임존 설정도 추가하였음

Mangcho committed:
추가 : graph 라우터 및 서비스 추가
기능은 구현하지 않았으나 차후 사용할 목적으로 파일 생성 및 기본적인 선언 입력

Mangcho committed:
추가, 보완 : pi, auth 라우터 입력값 검증 및 named export로 보완
1. pi라우터는 입력값 전체 null인지 검증기능 추가
2. 파이 백엔드 알고리즘 개발에 따른 result 컬럼 입력 추가
3. auth라우터 전체 입력값 검증 추가
4. 모든 라우터 named export로 변경

Mangcho committed:
보완, 수정 : auth, pi 서비스 변경
1. 파이 서비스 result 컬럼 추가
2. #7 지적 반영하여 timestamp 컬럼 대신 createdAt으로 변경
3. 리턴값 객체 대신 boolean으로 변경
4. 두 서비스 모두 named export로 변경

Mangcho committed:
추가, 보완 : named export 추가
named export에 따른 import문 변경
graphRouter 추가

Mangcho committed:
수정 : rasbpi 모델 컬럼명 수정
#7 코드리뷰 반영하여 timestamp에서 createdAt으로 변경

Mangcho committed:
수정 : whitelists 모델 옵션 변경
#7 코드리뷰 반영하여 updatedAt 생성

Mangcho committed:
추가 : 테스트 코드 추가
1. 처음 서버 시작시 자동으로 whitelists 및 users에 imei와 사용자 계정 추가

Mangcho committed:
보완 : graph 서비스 named export 적용

Mangcho committed:
보완 : getHash named export 적용

Mangcho committed:
보완, 추가 : auth 서비스 내 기능 추가 및 보완
1.getHash util named import
2.계정생성 Dto 변수명 변경
3.로그인 테스트 코드 삭제
4.비밀번호 변경 서비스 추가

Mangcho committed:
추가 : auth 라우트 기능 추가 및 보완
1. express-validator 헤더 검증 관련 import 추가
2. 비밀번호 변경 라우팅 추가
3. 로그아웃(세션 삭제) 관련 틀만 추가

Mangcho committed:
추가 : whitelist 테이블 id 컬럼 추가
기존 imei 컬럼은 우선키 대신 유니크로 변경

Mangcho committed:
수정 : user model의 변경에 따른 테스트 코드 컬럼명 변경

Mangcho committed:
추가 : uesr model 기본키 id 컬럼 추가
1. 자동으로 증가하는 uInt형 id 컬럼 추가
2. 기존에 id로 사용한 컬럼은 email로 변경

Mangcho committed:
수정 : auth 서비스 및 라우트 user 테이블 컬럼명 변경에 따른 변경
id에서 email로 변경

Mangcho committed:
수정 : 세션 관련 설정 변경
세션 사용하려고 이것저것 변경해보았음 보안상 위험

Mangcho committed:
추가 : 차후 그래프 기능 추가에 대비한 기본적인 양식 추가

Mangcho committed:
수정, 추가 : auth 서비스 코드 변경
1. 주석 수정
2. 로그인 서비스 세션 기능 추가
3. 로그아웃 서비스 기능 수정

Mangcho committed:
수정 : auth 라우터 기능 수정
1. 로그인 세션 내 isAuth 변수 삽입을 서비스로 옮김
2. 로그아웃 라우터 기능 추가 아직 제대로 작동하지 않음

Mangcho committed:
수정 : auth 라우터 로그아웃시 sid 서비스로 전달

Mangcho committed:
추가 : 세션 모델 신규 생성
1. 정상적으로 세션 인증 및 삭제가 되지 않는 관계로 자동 생성 대신 모델을 직접 구현하여 시퀄라이즈와 연결

Mangcho committed:
단순화 : 시퀄라이즈 config 단순한 문자열 수정

Mangcho committed:
단순화 : dotenv 로딩 파일인 env.js 권장사항대로 수정

Mangcho committed:
수정 : model index.js ENV 설정에 따라 세팅 변경하게 if문 사용

Mangcho committed:
수정 : Test 생성 코드 추가 및 세션 설정 변경
1. imei와 계정 생성하는 코드는 ENV 상태가 development 상태일 경우에만 사용하게 변경
2. connect-session-sequelize 옵션과 express-session 옵션의 위치를 변경하여 명확하게 보이도록함
3. 생성한 sessions 모델을 연결하여 직접 테이블 생성
4. 세션 쓸모없는 옵션 제거

Mangcho committed:
단순화 : sequelize.js 컨벤션 변경

Mangcho committed:
수정 : auth 라우터 로그아웃 기능 수정
1. 세션 정보를 서비스로 넘겨주게 수정

Mangcho committed:
단순화 : auth 서비스 로그아웃 기능 변수 정리
1. sessionID 변수가 정확하지 않고 해당 변수 없이도 작업이 가능하므로 삭제

Mangcho committed:
삭제 : ejs 모듈 삭제
어차피 사용하지 않았음

Mangcho committed:
단순화 : app.js 세션 옵션 위치 변경
1. 이전 커밋에서 확인하지 못한 위치를 제대로 다시 수정함
2. expires 옵션 관련하여 추가

Mangcho committed:
단순화 : auth 서비스에서 애러 확인용 console.log 삭제

kakasoo commented:
제가 제안한 것들은 기본적으로 필요를 느끼지 못한다면 안하셔도 되는 것들입니다. 코드 리뷰는 원래 그래요.

kakasoo commented:
직접 삭제해도 됩니다. 저는 세션 라이브러리를 써본 적 없어서 모릅니다만, 직접 세션을 구현하면서 저장소를 DB로 지정한다면 그런 형태가 될 거라고 생각합니다.

kakasoo commented:
에러는 콘솔만 찍는 게 아니고 수집합니다. 그리고 slow query는 따로 수집합니다. 에러가 아니더라도 너무 느린 응답은 캐치해둡니다. 제 서비스는 아직 규모가 작아서 저한테 메일이나 슬랙 알림을 주게 되어 있어요.

kakasoo commented on src/models/index.js:
```diff
@@ -3,12 +3,18 @@ const env = process.env;
 import { Sequelize, DataTypes } from 'sequelize';
 import {development, production } from "../settings/sequelize/config.js"
-const config = development;
+let config;
+if (env.NODE_ENV == 'development'){
+  config = development;
+} else {
+  config = production;
+}
```
const config = env.NODE_ENV === "development"? development: production

kakasoo replied:
let 사용 지양하시고,비교연산은 무조건 === 사용할 것.

kakasoo commented on src/routes/api/auth.js:
```diff
@@ -1,21 +1,91 @@
+      return res.json({ errors: errors.array() });
+    }
+    const compareUserDto = { email: req.body.email, password: req.body.password, session: req.session };
+    const state = await auth.SignIn(compareUserDto);
```
auth 대신 authService 라고 해주세요. 항상 레이어의 이름을 같이 씁시다. authRouter, authController, authService, authModel, authRepository 등등.

kakasoo commented on src/routes/api/auth.js:
```diff
@@ -1,21 +1,91 @@
+      age: req.body.age,
+      imei: req.body.imei,
+    };
+    const state = await auth.SignUp(createUserDto);
```
이미 존재하는 유저인지 체크하는 로직 필요.

kakasoo commented on src/services/auth.js:
```diff
@@ -1,49 +1,103 @@
+            // 객체를 보내주기에는 내용이 너무 많으므로 그냥 걸러주기
         } catch (error) {
-            console.log("SignUp Service error", error);
+            console.log("dafdas", error.errors);
+            console.log("SignUp Service", error);
```
이미 존재하는 유저일 경우 에러가 나오겠죠. email이 unique 하니깐요. 저는 이런 예측 가능한 에러는 비즈니스 로직에서 잡는 게 맞지, 에러로 취급할 게 아니라고 봐요. 에러는 개발자도 모르는 경우에 발생하는 거고, 그 마저도 잡으려고 해야죠.

kakasoo reviewed (approved):
라즈베리 부분은 아예 안봤습니다. 거기는 어차피 제가 도메인 지식이 없을 거라서요.

Mangcho committed:
추가 : graph 라우터, 서비스 기능 일부 구현
1. liveState 라우터 서비스랑 연결
2. liveState 서비스 쿼리문 추가

Mangcho committed:
추가 : test.js 테스트 코드 추가
1. 어소시에이션이 없는 테이블들에 대한 join 쿼리 구현
2. 시간을 조건으로 가지고 limit을 가진 쿼리문 구현

Mangcho committed:
수정 : auth 서비스 로그인 일부 변경
1. 쿼리 결과를 raw 옵션을 통해 일부 제한
2. 세션 데이터에 imei값 저장하게 추가

Mangcho committed:
수정 : app.js 일부 수정
1. 컨벤션 변경
2. 테스트코드 추가

Mangcho committed:
수정 : test.js 테스트 코드 수정
사소한 애러들 잡기 위해 사용하였음
추후 삭제 예정이므로 큰 의미 없음

Mangcho committed:
수정 : graph 라우터 척추신호등 기능 수정
1. 사용하지 않는 값들에 대한 검증 기능 삭제
2. 주석 삭제 및 Dto를 통한 서비스에서의 값 수신 및 반환 세팅

Mangcho committed:
수정 : graph 라우터 내 척추신호등 기능 수정
1. 쓸모없는 주석 삭제 및 함수 주석 변경
2. 쿼리 결과에 따라 값들의 평균 내고 이를 라우터로 반환하는 코드 추가

Mangcho committed:
수정 : graph 서비스 내 척추신호등 기능 수정
1. 좌우 척추 입력값의 평균을 따로 나눔. 좌우 합쳐서가 아닌 좌 따로 우따로
2. 경우에 따라 DB내 쿼리가 안될 경우가 있어서 쿼리가 안되었을경우 null 반환하게 설정

Mangcho committed:
단순화 : app.js 테스트코드 삭제

Mangcho committed:
수정 : auth라우터 코드 수정
1. #9 코드리뷰 지적 반영하여 클래스 인스턴스 명칭 변경 auth -> authService
2. 비밀번호 변경 내 Dto 요소 이름 변경

Mangcho committed:
수정 : graph라우터 코드 수정
1. #9 코드리뷰 지적 반영하여 클래스 인스턴스 명칭 변경 graph -> graphService
2. Dto내 삽입 데이터 변경

Mangcho committed:
단순화 : auth 서비스 내 비밀번호 변경 코드 단순화
1. 쿼리문 내 join연산 제거
2. where 절 내 변수 데이터 이름 변경

Mangcho committed:
수정 : #9 코드리뷰 지적 반영하여 model 내 index.js 변경
1. dev, production 모드 구분 코드 변경
2. 변수명 변경

Mangcho committed:
readme 수정

Mangcho committed:
추가 : package.json moment.js 설치 기록 추가됨

Mangcho committed:
추가 : sequelize UTC 타임에서 KST 한국시간으로 변경 코드 추가

Mangcho committed:
수정 : sequelize.js loader 코드 수정
프로젝트가 진행됨에 따라 force 옵션 대신 alter 옵션으로 데이터 보존

Mangcho committed:
추가 : graph 라우터 척추신호등, 사용기록 기능 추가
1. 세션에서 로그인한 사용자의 imei 값을 수신하여 서버에 전송
2. 전송후 척추신호등, 사용기록 서비스에 전달
3. 원본 센서 데이터 기록은 아직 미완성

Mangcho committed:
추가, 수정 : graph서비스 척추신호등 수정, 사용시간 기능 추가
1. 척추신호등 우측값의 합부분 버그 수정
2. 사용시간 서비스 기능 추가

Mangcho committed:
삭제 : app.js 테스트코드 삭제
